spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignClientAutoConfiguration.java

```diff
@@ -104,26 +104,26 @@ TraceFeignObjectWrapper traceFeignObjectWrapper(BeanFactory beanFactory) {
 
 	@Bean
 	@Primary
-	Decoder feignDecoder(BeanFactory beanFactory) {
-		return new TraceFeignDecoder(beanFactory,
-				new ResponseEntityDecoder(new SpringDecoder(this.messageConverters)) {
-					@Override
-					public Object decode(Response response, Type type)
-							throws IOException, FeignException {
-						FeignRequestContext feignRequestContext = FeignRequestContext
-								.getInstance();
-						FeignResponseHeadersHolder feignResponseHeadersHolder = new FeignResponseHeadersHolder(
-								response.headers());
-						feignResponseHeadersInjector().inject(
-								feignRequestContext.getCurrentSpan(),
-								feignResponseHeadersHolder);
-						return super.decode(
-								Response.create(response.status(), response.reason(),
-										feignResponseHeadersHolder.responseHeaders,
-										response.body()),
-								type);
+	Decoder feignDecoder(final BeanFactory beanFactory) {
+		return new ResponseEntityDecoder(new SpringDecoder(this.messageConverters)) {
+				@Override
+				public Object decode(Response response, Type type)
+						throws IOException, FeignException {
+					Tracer tracer = beanFactory.getBean(Tracer.class);
+					if (tracer == null) {
+						throw new IllegalStateException("Trace hasn't been yet initialized");
 					}
-				});
+					FeignResponseHeadersHolder feignResponseHeadersHolder = new FeignResponseHeadersHolder(
```

Comment:
Ah! It's this old approach where we injected trace stuff into the response. I think it should be deleted. Good catch!